File: Week 6/examples/change_making/dp_change.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> coins = {1, 5, 10, 25, 50}; // You can change this, and we will always get the right answer
int dp[400];
int min_coins(int n){
    int& ans = dp[n];
    if(ans != -1) return ans;
    ans = 1e9;
    for(int c : coins){
        // try using each coin
        if(n - c >= 0){
            ans = min(ans, 1 + min_coins(n - c));
        }
    }
    return ans;
}

vector<int> ans;
// call this after calling min_coins to get the actual coins
void get_coins(int n){
    if(n == 0) return;
    for(int c : coins){
        if(n - c >= 0 && dp[n] == 1 + dp[n - c]){
            ans.push_back(c);
            get_coins(n - c);
            return;
        }
    }
}

int main(){
    int n;
    cin >> n;
    memset(dp, -1, sizeof(dp));
    dp[0] = 0; // base case; 0 coins are needed to make 0
    cout << min_coins(n) << endl;
    get_coins(n);
    for(int i : ans) cout << i << " ";
    cout << endl;
}

File: Week 6/examples/change_making/dp_change.java
import java.util.*;
import java.io.*;

class Main{
    static int[] coins = {1, 5, 10, 25, 50}; // You can change this, and we will always get the right answer
    static int[] dp = new int[400];
    static int minCoins(int n){
        if(dp[n] != -1) return dp[n];
        int ans = Integer.MAX_VALUE;
        for(int c : coins){
            // try using each coin
            if(n - c >= 0){
                ans = Math.min(ans, 1 + minCoins(n - c));
            }
        }
        return dp[n] = ans;
    }
    static ArrayList<Integer> ans = new ArrayList<>();
    // call this after calling minCoins to get the actual coins
    static void getCoins(int n){
        if(n == 0) return;
        for(int c : coins){
            if(n - c >= 0 && dp[n] == 1 + dp[n - c]){
                ans.add(c);
                getCoins(n - c);
                return;
            }
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        Arrays.fill(dp, -1);
        dp[0] = 0; // base case; 0 coins are needed to make 0
        System.out.println(minCoins(n));
        getCoins(n);
        for(int i : ans) System.out.print(i + " ");
        System.out.println();
    }
}